adapters/akcelo/akcelo.go

```diff
@@ -0,0 +1,187 @@
+package akcelo
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/buger/jsonparser"
+	"github.com/prebid/openrtb/v20/openrtb2"
+	"github.com/prebid/prebid-server/v3/adapters"
+	"github.com/prebid/prebid-server/v3/config"
+	"github.com/prebid/prebid-server/v3/errortypes"
+	"github.com/prebid/prebid-server/v3/openrtb_ext"
+	"github.com/prebid/prebid-server/v3/util/jsonutil"
+	"github.com/prebid/prebid-server/v3/util/ptrutil"
+	"net/http"
+	"net/url"
+)
+
+type adapter struct {
+	uri *url.URL
+}
+
+type extObj struct {
+	Akcelo json.RawMessage `json:"akcelo"`
+}
+
+var noValidSiteIdError = &errortypes.BadInput{Message: "Cannot find valid siteId"}
+var noValidImpError = &errortypes.BadInput{Message: "No valid Imp"}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	uri, err := url.Parse(config.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+	bidder := &adapter{uri: uri}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	requestData, errs := a.prepareBidRequest(request)
+	return requestData, errs
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if adapters.IsResponseStatusCodeNoContent(response) {
+		return nil, nil
+	}
+	if err := adapters.CheckResponseStatusCodeForErrors(response); err != nil {
+		return nil, []error{err}
+	}
+	var bidResponse openrtb2.BidResponse
+	if err := jsonutil.Unmarshal(response.Body, &bidResponse); err != nil {
+		return nil, []error{err}
+	}
+	return extractBids(bidResponse)
+}
+
+func (a *adapter) prepareBidRequest(bidRequest *openrtb2.BidRequest) ([]*adapters.RequestData, []error) {
+	if len(bidRequest.Imp) == 0 {
+		return nil, []error{noValidImpError}
+	}
+
+	bidRequest = createSitePublisher(bidRequest)
+	requestData := make([]*adapters.RequestData, 0, 1)
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	var errs []error
+	for i := range bidRequest.Imp {
+		imp := &bidRequest.Imp[i]
+		err := a.prepareImp(imp)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+	}
+	if err := configureParentAccount(bidRequest); err != nil {
+		return nil, []error{err}
+	}
+
+	var resultBidRequest, err = jsonutil.Marshal(bidRequest)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	requestData = append(requestData, &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.uri.String(),
+		Body:    resultBidRequest,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(bidRequest.Imp),
+	})
+	return requestData, errs
+}
+
+func createSitePublisher(bidRequest *openrtb2.BidRequest) *openrtb2.BidRequest {
+	akceloRequest := *bidRequest
+	if akceloRequest.Site == nil {
```

Comment:
Up to this part of the code, a request must come with only one of these three:
 - `Site`
 - `App`
 - `DOOH`

Prebid Server filters out request that come with none of these or only one of them. Therefore, a request where `Site == nil` necessarily comes with either an `App` or `DOOH` non-nil field. If we're sticking with `Site`, should we set the other two, `App` or `DOOH`, to `nil` here? Or should we end up passing the `bidRequest.Imp[0].Ext.akcelo.siteId` value onto `bidRequest.App.Publisher.Ext` or `bidRequest.DOOH.Publisher.Ext`?